src/crypto/common_functions.cpp

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <libp2p/crypto/common_functions.hpp>
+
+#include <gsl/gsl_util>
+#include <libp2p/crypto/error.hpp>
+
+namespace libp2p::crypto {
+
+  outcome::result<std::shared_ptr<EC_KEY>> EcKeyFromPrivateKeyBytes(
+      int nid, gsl::span<const uint8_t> private_key) {
+    auto FAILED = KeyGeneratorError::INTERNAL_ERROR;
+
+    // try initialize requested curve group
+    EC_GROUP *curve = EC_GROUP_new_by_curve_name(nid);
+    if (nullptr == curve) {
+      return CryptoProviderError::UNKNOWN_KEY_TYPE;
+    }
+    auto free_curve = gsl::finally([curve] { EC_GROUP_free(curve); });
+
+    // allocate the key
+    std::shared_ptr<EC_KEY> key{EC_KEY_new(), EC_KEY_free};
```

Comment:
gsl::finally produces less code instructions and is more lightweight than sptr, but it is up to you.

Reply:
Thanks for the note, this is valuable! I tend to use shared pointers only for the cases when the initialized object comes out as a return value.